### PR TITLE
[AARCH64] Optimized addx()

### PR DESCRIPTION
**Pull request by aserna3**
Added some conditionals to check if one register is cached, but not both.

### CODE REVIEW

**Comment by Sonicadvance1** [comment]
Some code duplication can be removed by doing something like this paste.
http://hastebin.com/akotevutey.avrasm

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp [hunk 1]
```diff
@@ -282,8 +281,28 @@ void JitArm64::addx(UGeckoInstruction inst)
 		if (inst.Rc)
 			ComputeRC(gpr.GetImm(d), 0);
 	}
+	else if (gpr.IsImm(a) || gpr.IsImm(b))
+	{			
+        int imm_reg = gpr.IsImm(a) ? a : b;
```
Comment (Sonicadvance1) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp]: Whitespace needs to be tabs, not spaces

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp [hunk 2]
```diff
@@ -282,8 +281,28 @@ void JitArm64::addx(UGeckoInstruction inst)
...
+			MOVI2R(WA, gpr.GetImm(imm_reg));
+			ADD(gpr.R(d), gpr.R(in_reg), WA);
+			gpr.Unlock(WA);
+        }
+        if (inst.Rc)
+			ComputeRC(gpr.R(d), 0);
+	}
 	else
 	{
+
```
Comment (Sonicadvance1) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp]: There is an additional whitespace that shouldn't be here.

**Comment by degasus** [comment]
LGTM, but those whitespace issues.

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp [hunk 3]
```diff
@@ -273,14 +273,34 @@ void JitArm64::addx(UGeckoInstruction inst)
 	JITDISABLE(bJITIntegerOff);
 	FALLBACK_IF(inst.OE);
 
-	int a = inst.RA, b = inst.RB, d = inst.RD;
+        int a = inst.RA, b = inst.RB, d = inst.RD;
```
Comment (degasus) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp]: no whitespace

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp [hunk 4]
```diff
@@ -273,14 +273,34 @@ void JitArm64::addx(UGeckoInstruction inst)
...
 
 	if (gpr.IsImm(a) && gpr.IsImm(b))
 	{
 		s32 i = (s32)gpr.GetImm(a), j = (s32)gpr.GetImm(b);
 		gpr.SetImmediate(d, i + j);
 		if (inst.Rc)
-			ComputeRC(gpr.GetImm(d), 0);
+	                ComputeRC(gpr.GetImm(d), 0);
```
Comment (degasus) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp]: no whitespace

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp [hunk 5]
```diff
@@ -282,6 +282,26 @@ void JitArm64::addx(UGeckoInstruction inst)
...
+	else if (gpr.IsImm(a) || gpr.IsImm(b))
+	{                       
+		int imm_reg = gpr.IsImm(a) ? a : b;
+		int in_reg = gpr.IsImm(a) ? b : a;
+		gpr.BindToRegister(d, d == in_reg);         
+		if (gpr.GetImm(imm_reg) < 4096)
+		{
+			ADD(gpr.R(d), gpr.R(in_reg), gpr.GetImm(imm_reg));
+		}
+	else
```
Comment (Sonicadvance1) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp]: This needs to be indented one more

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp [hunk 6]
```diff
@@ -289,6 +309,8 @@ void JitArm64::addx(UGeckoInstruction inst)
 		if (inst.Rc)
 			ComputeRC(gpr.R(d), 0);
 	}
+
+}
```
Comment (Sonicadvance1) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp]: This enclosing bracket needs to be before the last else. Your indention is incorrect, so it is where the misreading is happening.

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp [hunk 7]
```diff
@@ -282,6 +282,26 @@ void JitArm64::addx(UGeckoInstruction inst)
 		if (inst.Rc)
 			ComputeRC(gpr.GetImm(d), 0);
 	}
+	else if (gpr.IsImm(a) || gpr.IsImm(b))
+	{                       
+		int imm_reg = gpr.IsImm(a) ? a : b;
+		int in_reg = gpr.IsImm(a) ? b : a;
+		gpr.BindToRegister(d, d == in_reg);
```
Comment (mathieui) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp]: Trailing whitespace

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp [hunk 8]
```diff
@@ -282,6 +282,26 @@ void JitArm64::addx(UGeckoInstruction inst)
 		if (inst.Rc)
 			ComputeRC(gpr.GetImm(d), 0);
 	}
+	else if (gpr.IsImm(a) || gpr.IsImm(b))
+	{
```
Comment (mathieui) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp]: Also trailing whitespace

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp [hunk 9]
```diff
@@ -282,12 +282,34 @@ void JitArm64::addx(UGeckoInstruction inst)
...
+			ARM64Reg WA = gpr.GetReg();
+			MOVI2R(WA, gpr.GetImm(imm_reg));
+			ADD(gpr.R(d), gpr.R(in_reg), WA);
+			gpr.Unlock(WA);
+		}
 		if (inst.Rc)
+		{
 			ComputeRC(gpr.R(d), 0);
+		}
+		else
```
Comment (Sonicadvance1) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp]: This else statement has to be outside of the 'if (imm(a) || imm(b)' if statement.

Comment (degasus) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp]: missing "}" before "else"

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp [hunk 10]
```diff
@@ -282,12 +282,34 @@ void JitArm64::addx(UGeckoInstruction inst)
...
+		{
+			gpr.BindToRegister(d, d == a || d == b);
+			ADD(gpr.R(d), gpr.R(a), gpr.R(b));
+			if (inst.Rc)
+				ComputeRC(gpr.R(d), 0);
+		}
+
 	}
```
Comment (degasus) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp]: This one should go to line 304

**Pull request merged**